### ISSUE TITLE
Conventional syntax for typeclass declarations and instances

### ISSUE THREAD

**Issue by dougalm** [google-research/dex-lang#185]
Right now, we define typeclasses by working with raw position-indexed dictionaries directly. For example (from the prelude):

```
data Add a:Type =
  MkAdd (a->a->a) (a->a->a) a  -- add, sub, zero

def (+)  (d:Add a) ?=> : a -> a -> a = case d of MkAdd add _   _    -> add
def (-)  (d:Add a) ?=> : a -> a -> a = case d of MkAdd _   sub _    -> sub
def zero (d:Add a) ?=> : a           = case d of MkAdd _   _   zero -> zero

@instance realAdd : Add Real = MkAdd (\x y. %fadd x y) (\x y. %fsub x y) 0.0
@instance intAdd  : Add Int  = MkAdd (\x y. %iadd x y) (\x y. %isub x y) 0
@instance unitAdd : Add Unit = MkAdd (\x y. ())        (\x y. ())        ()
```

Since @hexahedria added records in #180 , we should be able to give these dictionary entries names, and even add Haskell-like surface syntax for declarations and instances, `class ... where ...` and `instance ... where ...`.

**Comment by duvenaud** [comment]
A bit of bikeshedding:  When I first learned about these in Haskell, I found using ```data``` to denote something that's not data confusing.  I also found the word 'typeclass' confusing, and the fact that the keyword doesn't match the name of the thing.

Idris uses the word 'interface' for both, which I think is clearer:
https://idris2.readthedocs.io/en/latest/tutorial/interfaces.html

**Comment by gsp-27** [comment]
Hi,

I am extremely new to this, but I would like to a crack at it. Sorry for the dumb request, would it possible to point me to some resources to learn more about this.

Thanks,
Gaurav Pathak

**Comment by apaszke** [comment]
Implemented in #249.
